brainiak/matnormal/covs.py

```diff
@@ -3,438 +3,475 @@
-                                                           0]))
+            self.offdiag_template = tf.constant(
+                scipy.linalg.toeplitz(np.r_[0, 1, np.zeros(size - 2)]),
+                dtype=tf.float64
+            )
+            self.diag_template = tf.constant(np.diag(
+                  np.r_[0, np.ones(size - 2), 0]))
         else:
             self.run_sizes = np.ediff1d(np.r_[scan_onsets, size])
-            sub_offdiags = [scipy.linalg.toeplitz(np.r_[0, 1, np.zeros(r-2)])
-                            for r in self.run_sizes]
-            self.offdiag_template = tf.constant(scipy.sparse.
-                                                block_diag(sub_offdiags)
-                                                .toarray())
-            subdiags = [np.diag(np.r_[0, np.ones(r-2), 0])
+            sub_offdiags = [
+                scipy.linalg.toeplitz(np.r_[0, 1, np.zeros(r - 2)])
+                for r in self.run_sizes
+            ]
+            self.offdiag_template = tf.constant(
+                scipy.sparse.block_diag(sub_offdiags).toarray()
+            )
+            subdiags = [np.diag(np.r_[0, np.ones(r - 2), 0])
                         for r in self.run_sizes]
-            self.diag_template = tf.constant(scipy.sparse.
-                                             block_diag(subdiags)
-                                             .toarray())
+            self.diag_template = tf.constant(
+                scipy.sparse.block_diag(subdiags).toarray()
+            )
 
         self._identity_mat = tf.constant(np.eye(size))
 
         if sigma is None:
-            self.log_sigma = tf.Variable(tf.random_normal([1],
-                                         dtype=tf.float64), name="sigma")
+            self.log_sigma = tf.Variable(
+                tf.random.normal([1], dtype=tf.float64), name="sigma"
+            )
         else:
             self.log_sigma = tf.Variable(np.log(sigma), name="sigma")
 
         if rho is None:
-            self.rho_unc = tf.Variable(tf.random_normal([1], dtype=tf.float64),
-                                       name="rho")
+            self.rho_unc = tf.Variable(
+                tf.random.normal([1], dtype=tf.float64), name="rho"
+            )
         else:
-            self.rho_unc = tf.Variable(np.log(rho), name="rho")
+            self.rho_unc = tf.Variable(
+                2 * tf.sigmoid(self.rho_unc) - 1, name="rho")
 
-        super(CovAR1, self).__init__(size)
+    @property
+    def logdet(self):
+        """ log-determinant of this covariance
+        """
+        # first, unconstrain rho and sigma
+        rho = 2 * tf.sigmoid(self.rho_unc) - 1
+        sigma = tf.exp(self.log_sigma)
+        # now compute logdet
+        return tf.reduce_sum(
+            input_tensor=2 * tf.constant(self.run_sizes, dtype=tf.float64) *
+            tf.math.log(sigma)
+            - tf.math.log(1 - tf.square(rho))
+        )
```

Comment:
I think since you already have a self.log_sigma, you don't need to take exp and log again.